File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

function first_then_lower_case(strs, fxn)
  for _, str in ipairs(strs) do
    if fxn(str) then
      return string.lower(str)
    end
  end
  return nil
end

function powers_generator(base, limit)
  local power_coroutine = coroutine.create(function ()
    local power = 0
    while math.floor(base^power) <= limit do
      coroutine.yield(math.floor(base^power))
      power = power + 1
    end
  end)
  return power_coroutine
end

function say(words)
  local function inner(word)
    if word == nil then
      return words
    end
    return say(words .. " " .. word)
  end

  if words == nil then return "" end
  return inner
end

function meaningful_line_count(file)
  local f = io.open(file, "r")
  if f ==nil then
    error("No such file")
  end

  local lines = {}
  for line in io.lines(file) do
    line = line:gsub("^%s*", "")
    if line ~= nil and line ~= "" and string.sub(line, 1, 1) ~= "#" then
      table.insert(lines, line)
    end
  end

  f:close()
  return #lines
end

Quaternion = (function (class, meta, prototype)
  class.new = function (a, b, c, d)
    return setmetatable({a = a, b = b, c = c, d = d}, meta)
  end

  prototype.conjugate = function (self)
    return class.new(self.a, -self.b, -self.c, -self.d)
  end

  prototype.coefficients = function (self)
    return {self.a, self.b, self.c, self.d}
  end

  meta.__index = prototype

  meta.__add = function (self, q)
    return class.new(self.a + q.a, self.b + q.b, self.c + q.c, self.d + q.d)
  end

  meta.__mul = function(self, q)
    return class.new(
    (self.a * q.a - self.b * q.b - self.c * q.c - self.d * q.d),
    (self.a * q.b + self.b * q.a + self.c * q.d - self.d * q.c),
    (self.a * q.c - self.b * q.d + self.c * q.a + self.d * q.b),
    (self.a * q.d + self.b * q.c - self.c * q.b + self.d * q.a)
    )
  end

  meta.__eq = function (self, q)
    if self.a == q.a and self.b == q.b and self.c == q.c and self.d == q.d then
      return true
    else
      return false
    end
  end

  local function has_decimal (num)
    return num % 1 > 0
  end

  meta.__tostring = function (self)
    local return_str = ""
    if self.a == 0 and self.b == 0 and self.c == 0 and self.d == 0 then
      return_str = "0"
    end

    if self.a ~= 0 then
      if has_decimal(self.a) then
        return_str = return_str .. string.format("%g", self.a)
      else
        return_str = return_str .. string.format("%.1f", self.a)
      end
    end

    if self.b ~= 0 then
      if has_decimal(self.b) then
        return_str = return_str .. string.format("%+gi", self.b)
      else
        return_str = return_str .. string.format("%+.1fi", self.b)
      end
    end
    if math.abs(self.b) == 1 then return_str = return_str:gsub("1.0","") end

    if self.c ~= 0 then
      if has_decimal(self.c) then
        return_str = return_str .. string.format("%+gj", self.c)
      else
        return_str = return_str .. string.format("%+.1fj", self.c)
      end
    end
    if math.abs(self.c) == 1 then return_str = return_str:gsub("1.0","") end

    if self.d ~= 0 then
      if has_decimal(self.d) then
        return_str = return_str .. string.format("%+gk", self.d)
      else
        return_str = return_str .. string.format("%+.1fk", self.d)
      end
    end
    if math.abs(self.d) == 1 then return_str = return_str:gsub("1.0","") end

    if return_str:sub(1,1) == "+" then return_str = return_str:sub(2,#return_str) end

    return return_str
  end

  return class
end)({}, {}, {})

File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


def first_then_lower_case (strings, fxn, /):
    for str in strings:
        if fxn(str):
            return str.lower()

    return None

def powers_generator(*, base, limit):
    power = 0
    while pow(base, power) <= limit:
        yield pow(base, power)
        power = power + 1

def say(words=None, /):
    def inner(word=None, /):
        if word is None:
            return words
        return say(words + " " + word)

    if words is None:
        return ""
    return inner

def meaningful_line_count(file, /):
    f = open(file, encoding="utf8")
    if f is None:
        raise FileNotFoundError
    
    lines = []

    for line in f.readlines():
        line = line.strip()
        if line != None and line != "" and line[0] != "#":
            lines.append(line)

    f.close()
    return len(lines)

@dataclass(frozen=True)
class Quaternion:
    a: int
    b: int
    c: int
    d: int
    coefficients = tuple
    conjugate = tuple

    def __post__init__(self, /):
        object.__setattr__(self, 'a', self.a)
        object.__setattr__(self, 'b', self.b)
        object.__setattr__(self, 'c', self.c)
        object.__setattr__(self, 'd', self.d)

    @property
    def coefficients(self, /):
        return Quaternion(self.a, self.b, self.c, self.d)
    
    @property
    def conjugate(self, /):
        return Quaternion(self.a, -self.b, -self.c, -self.d)

    def __add__(self, q, /):
        return Quaternion(self.a + q.a, self.b + q.b, self.c + q.c, self.d + q.d)
    
    def __eq__(self, q, /):
        if type(q) == tuple and self.a == q[0] and self.b == q[1] and self.c == q[2] and self.d == q[3]:
            return True
        elif self.a == q.a and self.b == q.b and self.c == q.c and self.d == q.d:
            return True
        return False
    
    def __mul__(self, q, /):
        return Quaternion(
        (self.a * q.a - self.b * q.b - self.c * q.c - self.d * q.d),
        (self.a * q.b + self.b * q.a + self.c * q.d - self.d * q.c),
        (self.a * q.c - self.b * q.d + self.c * q.a + self.d * q.b),
        (self.a * q.d + self.b * q.c - self.c * q.b + self.d * q.a)
        )
    def has_decimal (self, num, /):
        return num % 1 > 0
    
    def __str__(self, /):
        return_str = ""
        if self.a == 0 and self.b == 0 and self.c == 0 and self.d == 0:
            return_str = "0"
            return return_str

        if self.a != 0:
            if self.has_decimal(self.a):
                return_str = return_str + f"{self.a}"
            else:
                return_str = return_str + f"{self.a:.1f}"

        if self.b != 0:
            if self.has_decimal(self.b):
                return_str = return_str + f"{self.b:+g}i"
            else:
                return_str = return_str + f"{self.b:+.1f}i"
        if abs(self.b) == 1:
            return_str = return_str.replace("1.0","")

        if self.c != 0:
            if self.has_decimal(self.c):
                return_str = return_str + f"{self.c:+g}j"
            else:
                return_str = return_str + f"{self.c:+.1f}j"
        if abs(self.c) == 1:
            return_str = return_str.replace("1.0","")

        if self.d != 0:
            if self.has_decimal(self.d):
                return_str = return_str + f"{self.d:+g}k"
            else:
                return_str = return_str + f"{self.d:+.1f}k"
        if abs(self.d) == 1:
            return_str = return_str.replace("1.0","")

        if return_str[0] == "+":
            return_str = return_str[1:]

        return return_str 
